python/llm/test/inference_gpu/test_optimize_model.py

```diff
@@ -54,6 +58,126 @@ def test_optimize_model(Model, Tokenizer, model_path):
     diff = abs(logits_base_model - logits_optimized_model).flatten()
 
     assert any(diff) is False
+"""
+
+class Test_Optimize_Gpu_Model:
+    def setup(self):
+
+        self.layer_inputs = []
+        self.layer_outputs = []
+        self.pre_layer_outputs = []
+
+    def run_optimize_gpu_model(self, Model, Tokenizer, model_path, self_attn, prev_attn, lower_bound):
+        def forward_hook(module, input, output, layer_name):
+            self.layer_outputs.append(output)
+
+        def pre_forward_hook(module, input, output, layer_name):
+            self.pre_layer_outputs.append(output)
+
+        def new_forward_hook(module, input):
+            if model_path == os.environ.get('LLAMA2_7B_ORIGIN_PATH'):
+                repalcement_norm = model.model.norm
+                opt_model.model.norm = repalcement_norm
+
+        def pre_hook(module, input):
+            self.layer_inputs.append(input)
+
+        def replace_forward_hook(module, input, output, layer_name):
+            hidden_states, present_key_value = output
+            hidden_states = layer_states[0].detach()
+            output = (hidden_states, present_key_value)
+            return output
+
+        tokenizer = Tokenizer.from_pretrained(model_path, trust_remote_code=True)
+        input_ids = tokenizer.encode(prompt, return_tensors="pt").to(device)
+
+        model = Model.from_pretrained(model_path,
+                                      load_in_4bit=True,
+                                      optimize_model=False,
+                                      trust_remote_code=True)
+        model = model.to(device)
+
+        for layer_name, layer_module in model.named_modules():
+            if layer_name == prev_attn:
+                layer_module.register_forward_pre_hook(
+                    lambda module, input: pre_hook(module, input))
+                layer_module.register_forward_hook(
+                    lambda module, input, output, layer_name=layer_name: pre_forward_hook(module, input,
```

Comment:
what's the purpose/benefit of adding this?

Reply:
to collect `hidden_states`, which is one of the output of DecoderLayer, this could be treat as the input for the next self_attn layer.